sys/include/configuration.h

```diff
@@ -0,0 +1,458 @@
+#else
+#define AUTO_INIT_CONFIGURATION(...)
+#endif
+
+#if !defined(CONFIG_CONFIGURATION_DEPTH_MAX) || defined(DOXYGEN)
+/**
+ * @brief   A path to a configuration item must not have more than this number
+ *          of segments
+ */
+#define CONFIG_CONFIGURATION_DEPTH_MAX      8
+#endif
+
+/**
+ * @brief   @ref CONFIG_CONFIGURATION_DEPTH_MAX
+ */
+#define CONFIGURATION_DEPTH_MAX             CONFIG_CONFIGURATION_DEPTH_MAX
+
+/**
+ * @brief   Forward declaration of a configuration handler
+ */
+struct conf_handler;
+
+/**
+ * @brief   Handler prototype to set a configuration value
+ *
+ * This is called by @ref configuration_set() for the handler.
+ * When @p val and @p size are NULL the configuration handler should delete the
+ * internal configuration value or mark it as deleted but not perform the
+ * backend delete operation.
+ *
+ * @param[in]           handler     Reference to the handler
+ * @param[in]           key         Configuration key which belongs to the configuration handler
+ * @param[in]           next        Comes after the key and the handler knows how to process
+ * @param[in]           val         Value to apply to the configuration item identified by the key
+ * @param[in, out]      size        Size in bytes of the value to set as input,
+ *                                  and 0 as output
+ *
+ * @return  0 on success
+ */
+typedef int (*conf_data_set_handler) (const struct conf_handler *handler,
+                                      char *key, char *next, const void *val,
+                                      size_t *size);
+
+/**
+ * @brief   Handler prototype to get a configuration value
+ *
+ * This is called by @ref configuration_get() for the handler.
+ *
+ * @param[in]           handler     Reference to the handler
+ * @param[in]           key         Configuration key which belongs to the configuration handler
+ * @param[in]           next        Comes after the key and the handler knows how to process
+ * @param[in]           val         Destination to write the configuration item to
+ * @param[in, out]      size        Maximum size in bytes of the value get as input,
+ *                                  and remaining length as output
+ *
+ * @return  0 on success
+ */
+typedef int (*conf_data_get_handler) (const struct conf_handler *handler,
+                                      char *key, char *next, void *val,
+                                      size_t *size);
+
+/**
+ * @brief   Handler prototype to import a configuration value from persistent storage to
+ *          the internal representation location of the configuration item
+ *
+ * This is called by @ref configuration_import() for the handler.
+ *
+ * @param[in]           handler     Reference to the handler
+ * @param[in]           key         Configuration key which belongs to the configuration handler
+ * @param[in]           next        Comes after the key and the handler knows how to process
```

Comment:
This I don't understand

Reply:
If `key` in `configuration_set()` is `foo/bar/bazz/length` and `foo`, `bar`, `bazz` are nodes in the tree where `bazz` implements a handler, `key` in the handler callback would be `foo/bar/bazz` and `next` would be `length`.
Also we could maybe reach the and of the tree already by `/foo/bar` in which case, `next` would be `bazz/length`.
I would maybe assume that `/foo/bar/bazz` would be the key in persistent storage to import/export the full object and `foo/bar/bazz/length` could be used to just set/get the length of bazz.